lib/ansible/executor/task_executor.py

```diff
@@ -80,6 +80,31 @@ def __init__(self, host, task, job_vars, play_context, new_stdin, loader, shared
         self._final_q = final_q
         self._loop_eval_error = None
 
+        from collections import defaultdict
+        from ansible.module_utils._text import to_text, to_bytes, to_native
+        from yaml import load as yaml_load
+        try:
+            # use C version if possible for speedup
+            from yaml import CSafeLoader as SafeLoader
+        except ImportError:
+            from yaml import SafeLoader
+        mod_defs_file = to_bytes(C.config.get_config_value('MODULE_DEFAULTS_CFG') or '%s/module_defaults.yml' % os.path.join(os.path.dirname(__file__), '..', 'config'))
+        if os.path.exists(mod_defs_file):
+            with open(mod_defs_file, 'rb') as config_def:
+                module_default_groups = yaml_load(config_def, Loader=SafeLoader).get('groups', {})
+        else:
+            raise AnsibleError("Missing base module_defaults definition file (bad install?): %s" % to_native(mod_defs_file))
+        if C.config.get_config_value('MODULE_DEFAULTS_CFG') is not None and os.path.exists(to_bytes(C.config.get_config_value('MODULE_DEFAULTS_CFG'))):
+            with open(to_bytes(C.config.get_config_value('MODULE_DEFAULTS_CFG')), 'rb') as config_def:
+                module_default_groups = module_default_groups.update(yaml_load(config_def, Loader=SafeLoader).get('groups', {}))
+        elif C.config.get_config_value('MODULE_DEFAULTS_CFG') is not None:
+            raise AnsibleError("Missing user-specified MODULE_DEFAULTS_CFG file: %s" % to_native(C.config.get_config_value('MODULE_DEFAULTS_CFG')))
+        m_to_g = dict()
+        for group, modules in module_default_groups.items():
+            for module in modules:
+                m_to_g[module] = group
+        self._module_to_group_map = m_to_g
```

Comment:
Does this handle modules that exist in multiple groups?

Reply:
Currently it doesn't, but I don't quite know the right way to account for that, since I hadn't thought about a precedence system. For example: 

```
module_defaults:
  aws_s3_bucket_facts:
    region: us-west-2
  group/aws:
    region: us-east-1
```

Has a clear precedence heirarchy, because the more specific non-group version should win because it's a closer scope than just the group. If you had, for example, two groups like `group/azure` and `group/azure-stack` or something, which one should take precedence isn't clear. I think this is only an issue if the user both uses a module in >1 group and also specifies identical parameters on those groups. Allowing users to specify multiple non-overlapping parameters for modules that share groups like this:

```
module_defaults:
  group/azure:
    key: ......
  group/azure_stack:
    region: dc-local
```

Would be relatively easy to implement, but I'm not sure what a users intention would be with multiple groups and defining the same parameters on those groups. 
Would be easy to